biocrnpyler/dna_construct.py

```diff
@@ -397,12 +403,142 @@ def enumerate_components(self,previously_enumerated=None):
+                        best_partlist += [(next_parthash,next_part)]
+                if(test_partlist[-1][0] > best_partlist[-1][0]):
+                    best_partlist = test_partlist
+        while(len(best_partlist)<len(construct)):
+            nextpart = circular_next(best_partlist[-1][1],construct)
+            best_partlist += [(Construct.get_partstring(nextpart),nextpart)]
+        if(return_rotation):
+            return Construct.get_partlist_hash(best_partlist),best_partlist[0][1].position
+        else:
+            return Construct.get_partlist_hash(best_partlist)
+    @classmethod
+    def direction_rotation_free_hash(cls,construct,return_transform = False):
+        """computes the best circular permutation of a construct, forward and reverse. Then returns whichever one of those comes alphabetically first"""
+        rev_con = Construct.hashless_reverse(construct)
+        forward_hash,pos = Construct.rotation_free_hash(construct,return_rotation=True)
+        reverse_hash,posrev = Construct.rotation_free_hash(rev_con,return_rotation=True)
+        if(forward_hash>reverse_hash):
+            if(return_transform):
+                return forward_hash,1,pos
+            else:
+                return forward_hash
+        else:
+            if(return_transform):
+                return reverse_hash,-1,posrev
+            else:
+                return reverse_hash
+    @classmethod
+    def linear_direction_free_hash(cls,construct,return_transform = False):
+        """creates a string representing the construct forward or reverse, and returns whichever of those is alphabetically first"""
+        rev_con = Construct.hashless_reverse(construct)
+
+        test_partlist = []
+        test_reverse_partlist = []
+        winner = None
+        for part,revpart in zip(construct,rev_con):
+            if(winner is None):
+                parthash = Construct.get_partstring(part)
+                rev_parthash = Construct.get_partstring(revpart)
+                if(parthash>rev_parthash):
+                    winner = "forward"
+                elif(parthash<rev_parthash):
+                    winner = "reverse"
+                test_partlist += [(parthash,part)]
+                test_reverse_partlist += [(rev_parthash,revpart)]
+            elif(winner=="forward"):
+                test_partlist += [(Construct.get_partstring(part),part)]
+            elif(winner=="reverse"):
+                test_reverse_partlist += [(Construct.get_partstring(revpart),revpart)]
+        if(winner=="forward"):
+            rethash = Construct.get_partlist_hash(test_partlist)
+            if(return_transform):
+                return rethash,1,0
+            return rethash
+        elif(winner == "reverse"):
+            rethash = Construct.get_partlist_hash(test_reverse_partlist)
+            if(return_transform):
+                return rethash,-1,0
+            return rethash
+    @classmethod
+    def omnihash(cls,construct,return_transform=False):
```

Comment:
This is the most important as this is the interface to the hashing part. Can you remove the `return_transform=True` branch of the code? Otherwise, we need to document the return type and create tests for it.